File: Greedy/Leetcode/Medium/FractionalKnapsack/FractionalKnapsack.cpp
// https://www.geeksforgeeks.org/problems/fractional-knapsack-1587115620/1
#include<bits/stdc++.h>
using namespace std;
double fractionalKnapsack(vector<int> &values, vector<int> &weights, int w)
{
    // Your code here
    vector<pair<int, int>> p;
    for (int i = 0; i < values.size(); i++)
    {
        p.emplace_back(make_pair(values[i], weights[i]));
    }

    sort(p.begin(), p.end(), [](pair<int, int> &a, pair<int, int> &b)
         { return (double)b.first / b.second < (double)a.first / a.second; });

    double val = 0.0;
    int i = 0;
    while (i < values.size() && w != 0)
    {
        int value = p[i].first, wt = p[i].second;
        if (w - wt >= 0)
        {
            w = w - wt;
            val += value;
        }
        else
        {
            double vpw = (double)value / wt;
            val += w * vpw;
            w = 0;
        }
        i++;
    }
    return val;
}
int main(void) {
    
    return 0;
}

File: Greedy/GFG/Medium/FractionalKnapsack/FractionalKnapsack.java
// https://www.geeksforgeeks.org/problems/fractional-knapsack-1587115620/1

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Pair {

    int val, wt;

    Pair(int val, int wt) {
        this.val = val;
        this.wt = wt;
    }
}
public class FractionalKnapsack {
    double fractionalKnapsack(List<Integer> values, List<Integer> weights, int w) {
        // code here
        List<Pair> p = new ArrayList<>();
        for(int i=0; i<values.size(); i++) {
            p.add(new Pair(values.get(i), weights.get(i)));
        }
        Collections.sort(p, (a, b) -> Double.compare((double)b.val / b.wt, (double)a.val / a.wt));
        double value = 0.0;
        int i=0;
        while(i < values.size() && w != 0) {
            int weight = p.get(i).wt, v = p.get(i).val;
            if(w - weight >= 0) {
                w -= weight;
                value += v;
            } else {
                double vpw = (double)v/weight;
                value += w*vpw;
                w = 0;
            }
            i++;
        }
        return value;
    }
}
